votor/src/certificate_pool/slot_stake_counters.rs

```diff
@@ -0,0 +1,301 @@
+                self.notarize_total = self.notarize_entry_total.values().sum();
+                self.top_notarized_stake = self.top_notarized_stake.max(entry_stake);
+            }
+            _ => return, // Not interested in other vote types
+        }
+        if self.my_first_vote.is_none() && is_my_own_vote {
+            self.my_first_vote = Some(*vote);
+            self.total_stake = total_stake;
+        }
+        if self.my_first_vote.is_some() {
+            let slot = vote.slot();
+            // Check safe to notar
+            for (block_id, stake) in &self.notarize_entry_total {
+                if !self.safe_to_notar_sent.contains(block_id)
+                    && self.is_safe_to_notar(block_id, stake)
+                {
+                    events.push(VotorEvent::SafeToNotar((slot, *block_id)));
+                    stats.event_safe_to_notarize = stats.event_safe_to_notarize.saturating_add(1);
+                    self.safe_to_notar_sent.push(*block_id);
+                }
+            }
+            // Check safe to skip
+            if !self.safe_to_skip_sent && self.is_safe_to_skip() {
+                events.push(VotorEvent::SafeToSkip(slot));
+                self.safe_to_skip_sent = true;
+                stats.event_safe_to_skip = stats.event_safe_to_skip.saturating_add(1);
+            }
+        }
+    }
+
+    fn is_safe_to_notar(&self, block_id: &Hash, stake: &Stake) -> bool {
+        // White paper v1.1 page 22: The event is only issued if the node voted in slot s already,
+        // but not to notarize b. Moreover:
+        // notar(b) >= 40% or (skip(s) + notar(b) >= 60% and notar(b) >= 20%)
+        if let Some(Vote::Notarize(my_vote)) = self.my_first_vote.as_ref() {
+            if my_vote.block_id() == block_id {
+                return false; // I voted for the same block, no need to send NotarizeFallback
+            }
+        }
+        let skip_ratio = self.skip_total as f64 / self.total_stake as f64;
+        let notarized_ratio = *stake as f64 / self.total_stake as f64;
+        // Check if the block fits condition (i) 40% of stake holders voted notarize
+        notarized_ratio >= SAFE_TO_NOTAR_MIN_NOTARIZE_ONLY
+            // Check if the block fits condition (ii) 20% notarized, and 60% notarized or skip
+            || (notarized_ratio >= SAFE_TO_NOTAR_MIN_NOTARIZE_FOR_NOTARIZE_OR_SKIP
+                && notarized_ratio + skip_ratio >= SAFE_TO_NOTAR_MIN_NOTARIZE_AND_SKIP)
+    }
+
+    fn is_safe_to_skip(&self) -> bool {
+        // White paper v1.1 page 22: The event is only issued if the node voted in slot s already,
+        // but not to skip s. Moreover:
+        // skip(s) + Sum of all notarize - (max in notarize(b)) >= 40%
+        if let Some(Vote::Notarize(_)) = self.my_first_vote.as_ref() {
+            error!(
+                "{} {} {}",
+                self.skip_total, self.notarize_total, self.top_notarized_stake
+            );
```

Comment:
might be interesting to log the voted block id here as well

Reply:
Oops, this was for debugging, changed to trace!